tests/integration/test_charm.py

```diff
@@ -176,11 +181,98 @@ async def test_monitor_user(ops_test: OpsTest) -> None:
+    password1 = await get_password(ops_test, username="monitor")
+    assert password1 != password
+
+    await set_password(ops_test, unit_id=leader_id, username="monitor", password=password)
+    for _ in UNIT_IDS:
+        password2 = await get_password(ops_test, username="monitor")
+        assert password2 == password
+
+
+@pytest.mark.usefixtures("only_with_juju_secrets")
+async def test_reset_and_get_password_secret_same_as_cli(ops_test: OpsTest) -> None:
+    """Test verifies that we can set and retrieve the correct password using Juju 3.x secrets."""
+    new_password = str(uuid4())
+
+    # Re=setting existing password
+    leader_id = await get_leader_id(ops_test)
+    result = await set_password(
+        ops_test, unit_id=leader_id, username="monitor", password=new_password
+    )
+
+    secret_id = result["secret-id"].split("/")[-1]
+
+    # Getting back the pw programmatically
+    password = await get_password(ops_test, username="monitor")
+
+    # Getting back the pw from juju CLI
+    complete_command = f"show-secret {secret_id} --reveal --format=json"
+    _, stdout, _ = await ops_test.juju(*complete_command.split())
+    data = json.loads(stdout)
+
+    assert password == new_password
+    assert data[secret_id]["content"]["Data"]["monitor-password"] == password
+
+
+@pytest.mark.usefixtures("only_without_juju_secrets")
+async def test_reset_and_get_password_no_secret(ops_test: OpsTest, mocker) -> None:
+    """Test verifies that we can set and retrieve the correct password using Juju 2.x."""
+    new_password = str(uuid4())
+
+    # Re=setting existing password
+    leader_id = await get_leader_id(ops_test)
+    await set_password(ops_test, unit_id=leader_id, username="monitor", password=new_password)
+
+    # Getting back the pw programmatically
+    password = await get_password(ops_test, username="monitor")
+    assert password == new_password
+
+
+@pytest.mark.usefixtures("only_with_juju_secrets")
+async def test_empty_password(ops_test: OpsTest) -> None:
+    """Test that the password can't be set to an empty string."""
+    leader_id = await get_leader_id(ops_test)
+
+    password1 = await get_password(ops_test, username="monitor")
+    await set_password(ops_test, unit_id=leader_id, username="monitor", password="")
+    password2 = await get_password(ops_test, username="monitor")
+
+    # The password remained unchanged
+    assert password1 == password2
+
+
+@pytest.mark.usefixtures("only_with_juju_secrets")
+async def test_no_password_change_on_invalid_password(ops_test: OpsTest) -> None:
+    """Test that in general, there is no change when password validation fails."""
+    leader_id = await get_leader_id(ops_test)
+    password1 = await get_password(ops_test, username="monitor")
+
+    # The password has to be minimum 3 characters
+    await set_password(ops_test, unit_id=leader_id, username="monitor", password="ca" * 1000000)
```

Comment:
`* 1000000)`?

Reply:
This test was taken from K8S charm as is :) 

And actually helped with setting max password length (otherwise there will be a crash in setting params for snap in connect_to_exporter)

Reply:
Yes, this was my purposefully invalid password, passing the limit :-)  * 1000000 :-)